### PR TITLE
Represent browsers as objects internally

### LINKED ISSUE #3183: Chromium installed via snap is not available in Cypress
### Current behavior:

If Chromium is installed as a snap (`snap install chromium`), Cypress won't auto-detect it as a supported browser for testing:

image

### Desired behavior:

Make Chromium installed as a snap available for testing without additional configuration, like the docs say: "[Cypress automatically detects available browsers on your OS.](https://docs.cypress.io/guides/guides/launching-browsers.html#Browsers)"

### Steps to reproduce: (app code and test code)

1. Install Chromium as a snap
2. Open the Cypress GUI
3. Try to use it as a browser

### Versions

Cypress: latest develop branch  
Ubuntu 18.04.1  
Chromium 71 stable from https://snapcraft.io/chromium

### Potentially related issues

#1400

### LINKED ISSUE #3218: Represent browsers as `Browser` objects internally
### Current behavior:

Inside the launcher package, browser is treated as an object with a name, a version, a path to the browser...

[packages/launcher/lib/types.ts](https://github.com/cypress-io/cypress/blob/develop/packages/launcher/lib/types.ts#L6)

Inside the server package, the browser is treated as an object (at first), but is quickly reduced to just the name (`chrome`, `chromium`, ...)

This means there's not much flexibility in how we can detect different browsers and then pass that information to the rest of Cypress.

### Desired behavior:

Modify the internal representation of browsers from just the name to always be the full Browser object, to enable new functionality like selecting browser by version (#3183), using arbitrary paths to browsers (#1026), and using proper display names in the desktop-gui.

### PR DESCRIPTION
**Pull request by flotwig**
Fixes #3218 
Fixes #3183

### CODE REVIEW

File: packages/server/lib/browsers/index.coffee [hunk 1]
```diff
@@ -32,30 +32,24 @@ cleanup = ->
 
 getBrowser = (name) ->
   switch name
-    ## normalize all the chrome* browsers
-    when "chrome", "chromium", "canary"
-      require("./chrome")
     when "electron"
       require("./electron")
+    else
+      require("./chrome")
```
Comment (brian-mann) [inline packages/server/lib/browsers/index.coffee]: whenever we add more browsers this will have to change back to what it was

Comment (flotwig) [reply packages/server/lib/browsers/index.coffee]: yep. made it fall-through for now so we can add more editions of Chrome without having to edit this file - `chrome-beta` is the one I'm thinking of in particular

File: packages/server/lib/browsers/utils.coffee [hunk 2]
```diff
@@ -7,6 +7,12 @@ profileCleaner = require("../util/profile_cleaner")
 
 PATH_TO_BROWSERS = appData.path("browsers")
 
+getBrowserPath = (browser) ->
```
Comment (brian-mann) [inline packages/server/lib/browsers/utils.coffee]: after investigating, there's no reason to namespace this further based on the way we handle profile creation and partitions

File: packages/server/lib/browsers/utils.coffee [hunk 3]
```diff
@@ -86,7 +92,6 @@ module.exports = {
         info: "Electron is the default browser that comes with Cypress. This is the browser that runs in headless mode. Selecting this browser is useful when debugging. The version number indicates the underlying Chromium version that Electron uses."
       })
 
-  launch: (name, url, args) ->
-    launcher()
-    .call("launch", name, url, args)
+  launch: (browser, url, args) ->
```
Comment (brian-mann) [inline packages/server/lib/browsers/utils.coffee]: `launch: launcher.launch`

**Pull request merged**